File: js/orbit.js
/* Traza terrestre de una órbita alrededor de Kerbin.

   KSP usa cónicas parcheadas: dentro de la SOI de Kerbin la órbita es una
   elipse kepleriana exacta, sin achatamiento ni J2 que la perturben. Así que
   basta propagar Kepler y restar la rotación del planeta para pasar del marco
   inercial al fijo al cuerpo. */

(function () {
  const B = KM.BODY;
  const D2R = Math.PI / 180, R2D = 180 / Math.PI;

  /* Ecuación de Kepler M = E - e·sen E, por Newton-Raphson. */
  function eccentricAnomaly(M, e) {
    M = ((M % (2 * Math.PI)) + 2 * Math.PI) % (2 * Math.PI);
    let E = e < 0.8 ? M : Math.PI;
    for (let i = 0; i < 60; i++) {
      const f = E - e * Math.sin(E) - M;
      const fp = 1 - e * Math.cos(E);
      const d = f / fp;
      E -= d;
      if (Math.abs(d) < 1e-12) break;
    }
    return E;
  }

  KM.orbit = {
    /* pe/ap en metros sobre el nivel del mar; ángulos en grados. */
    compute(opts) {
      let pe = opts.pe, ap = opts.ap;
      if (ap < pe) { const t = ap; ap = pe; pe = t; }

      const rp = B.radius + pe, ra = B.radius + ap;
      const a = (rp + ra) / 2;
      const e = (ra - rp) / (ra + rp);
      const n = Math.sqrt(B.mu / (a * a * a));     // movimiento medio, rad/s
      const T = 2 * Math.PI / n;                   // periodo orbital, s
      const wb = 2 * Math.PI / B.siderealDay;      // rotación de Kerbin, rad/s

      const inc = opts.inc * D2R, lan = opts.lan * D2R, argp = opts.argp * D2R;
      const orbits = Math.max(1, Math.min(60, opts.orbits | 0 || 1));

      /* Más muestras en órbitas excéntricas: cerca del periapsis la traza
         avanza muy rápido y con pocos puntos se ve angulosa. */
      const perOrbit = Math.round(180 * (1 + 2 * e));
      const steps = orbits * perOrbit;
      const pts = [];

      for (let i = 0; i <= steps; i++) {
        const t = (i / steps) * orbits * T;
        const E = eccentricAnomaly(n * t, e);
        const nu = 2 * Math.atan2(Math.sqrt(1 + e) * Math.sin(E / 2),
                                  Math.sqrt(1 - e) * Math.cos(E / 2));
        const r = a * (1 - e * Math.cos(E));
        const u = argp + nu;                        // argumento de latitud

        const X = r * (Math.cos(lan) * Math.cos(u) - Math.sin(lan) * Math.sin(u) * Math.cos(inc));
        const Y = r * (Math.sin(lan) * Math.cos(u) + Math.cos(lan) * Math.sin(u) * Math.cos(inc));
        const Z = r * (Math.sin(u) * Math.sin(inc));

        const lat = Math.asin(Math.max(-1, Math.min(1, Z / r))) * R2D;
        const lon = KM.geo.wrapLon((Math.atan2(Y, X) - wb * t) * R2D);

        pts.push({ lat, lon, alt: r - B.radius, t });
      }

      const vPe = Math.sqrt(B.mu * (2 / rp - 1 / a));
      const vAp = Math.sqrt(B.mu * (2 / ra - 1 / a));
      const drift = -(T / B.siderealDay) * 360;     // desplazamiento de la traza por vuelta

      return {
        points: pts,
        segments: KM.geo.splitAntimeridian(pts.map(p => [p.lat, p.lon])),
        a, e, T, vPe, vAp, drift,
        pe, ap,
        maxLat: Math.min(90, Math.abs(opts.inc) <= 90 ? Math.abs(opts.inc) : 180 - Math.abs(opts.inc)),
        footprintPe: KM.geo.horizonRadius(pe),
        footprintAp: KM.geo.horizonRadius(ap),
        synchronous: Math.abs(T - B.siderealDay) / B.siderealDay < 0.005,
        suborbital: pe < 0,
        inAtmosphere: pe < B.atmosphere
      };
    }
  };
})();

File: desktop/src/Core/Orbit.cs
using System;
using System.Collections.Generic;

namespace KerbinMaps.Core
{
    public readonly record struct TrackPoint(double Lat, double Lon, double Alt, double T);

    public sealed class OrbitResult
    {
        public List<TrackPoint> Points;
        public double A, E, T, VPe, VAp, Drift, Pe, Ap, MaxLat, FootprintPe, FootprintAp;
        public bool Synchronous, Suborbital, InAtmosphere;
    }

    /* Traza terrestre de una órbita alrededor de Kerbin.

       KSP usa cónicas parcheadas: dentro de la SOI de Kerbin la órbita es una elipse
       kepleriana exacta, sin achatamiento ni J2. Basta propagar Kepler y restar la
       rotación del planeta para pasar del marco inercial al fijo al cuerpo. */
    public static class ManualOrbit
    {
        /* Ecuación de Kepler M = E - e·sen E, por Newton-Raphson. */
        public static double EccentricAnomaly(double M, double e, double tol = 1e-12)
        {
            M = ((M % (2 * Math.PI)) + 2 * Math.PI) % (2 * Math.PI);
            double E = e < 0.8 ? M : Math.PI;
            for (int i = 0; i < 60; i++)
            {
                double d = (E - e * Math.Sin(E) - M) / (1 - e * Math.Cos(E));
                E -= d;
                if (Math.Abs(d) < tol) break;
            }
            return E;
        }

        /* pe/ap en metros sobre el nivel del mar; ángulos en grados. */
        public static OrbitResult Compute(double pe, double ap, double incDeg, double lanDeg, double argpDeg, int orbitsIn)
        {
            if (ap < pe) (pe, ap) = (ap, pe);

            double rp = Body.Radius + pe, ra = Body.Radius + ap;
            double a = (rp + ra) / 2;
            double e = (ra - rp) / (ra + rp);
            double n = Math.Sqrt(Body.Mu / (a * a * a));     // movimiento medio, rad/s
            double T = 2 * Math.PI / n;                       // periodo orbital, s
            double wb = 2 * Math.PI / Body.SiderealDay;       // rotación de Kerbin, rad/s

            double inc = incDeg * Geo.D2R, lan = lanDeg * Geo.D2R, argp = argpDeg * Geo.D2R;
            int orbits = Math.Max(1, Math.Min(60, orbitsIn == 0 ? 1 : orbitsIn));

            /* Más muestras en órbitas excéntricas: cerca del periapsis la traza avanza
               muy rápido y con pocos puntos se ve angulosa. */
            int perOrbit = (int)Math.Round(180 * (1 + 2 * e), MidpointRounding.AwayFromZero);
            int steps = orbits * perOrbit;
            var pts = new List<TrackPoint>(steps + 1);

            for (int i = 0; i <= steps; i++)
            {
                double t = ((double)i / steps) * orbits * T;
                double E = EccentricAnomaly(n * t, e);
                double nu = 2 * Math.Atan2(Math.Sqrt(1 + e) * Math.Sin(E / 2), Math.Sqrt(1 - e) * Math.Cos(E / 2));
                double r = a * (1 - e * Math.Cos(E));
                double u = argp + nu;                          // argumento de latitud

                double X = r * (Math.Cos(lan) * Math.Cos(u) - Math.Sin(lan) * Math.Sin(u) * Math.Cos(inc));
                double Y = r * (Math.Sin(lan) * Math.Cos(u) + Math.Cos(lan) * Math.Sin(u) * Math.Cos(inc));
                double Z = r * (Math.Sin(u) * Math.Sin(inc));

                double lat = Math.Asin(Math.Max(-1, Math.Min(1, Z / r))) * Geo.R2D;
                double lon = Geo.WrapLon((Math.Atan2(Y, X) - wb * t) * Geo.R2D);
                pts.Add(new TrackPoint(lat, lon, r - Body.Radius, t));
            }

            return new OrbitResult
            {
                Points = pts,
                A = a, E = e, T = T,
                VPe = Math.Sqrt(Body.Mu * (2 / rp - 1 / a)),
                VAp = Math.Sqrt(Body.Mu * (2 / ra - 1 / a)),
                Drift = -(T / Body.SiderealDay) * 360,        // desplazamiento de la traza por vuelta
                Pe = pe, Ap = ap,
                MaxLat = Math.Min(90, Math.Abs(incDeg) <= 90 ? Math.Abs(incDeg) : 180 - Math.Abs(incDeg)),
                FootprintPe = Geo.HorizonRadius(pe),
                FootprintAp = Geo.HorizonRadius(ap),
                Synchronous = Math.Abs(T - Body.SiderealDay) / Body.SiderealDay < 0.005,
                Suborbital = pe < 0,
                InAtmosphere = pe < Body.Atmosphere
            };
        }
    }
}
